Assets/Scripts/StandardSamples/Services/PlayerDataStorageService.cs

```diff
@@ -0,0 +1,468 @@
+                await _playerDataStorageSemaphore.WaitAsync();
+
+                QueryFileListOptions options = new QueryFileListOptions
+                {
+                    LocalUserId = userId
+                };
+
+                // Query the file list, and on callback set the TaskCompletionSource's result if possible
+                EOSManager.Instance.GetPlayerDataStorageInterface().QueryFileList(ref options, null, (ref QueryFileListCallbackInfo callbackInfo) =>
+                {
+                    try
+                    {
+                        // If there are no files for this user, the query will return "not found", so give an empty list
+                        if (callbackInfo.ResultCode == Result.NotFound)
+                        {
+                            tcs.SetResult(new List<string>());
+                            return;
+                        }
+
+                        if (callbackInfo.ResultCode != Result.Success)
+                        {
+                            tcs.SetException(new Exception($"{nameof(PlayerDataStorageService)} ({nameof(QueryFileList)}): Failed to query file list. Result code {callbackInfo.ResultCode}"));
+                            return;
+                        }
+
+                        uint fileCount = callbackInfo.FileCount;
+                        PlayerDataStorageInterface playerStorageHandle = EOSManager.Instance.GetPlayerDataStorageInterface();
+                        List<string> fileNames = new List<string>();
+
+                        // With the retrieved files, get the name of each file (which is in the metadata),
+                        // and compose a list of the file names
+                        for (uint fileIndex = 0; fileIndex < callbackInfo.FileCount; fileIndex++)
+                        {
+                            CopyFileMetadataAtIndexOptions options = new CopyFileMetadataAtIndexOptions()
+                            {
+                                LocalUserId = userId,
+                                Index = fileIndex
+                            };
+
+                            Result result = playerStorageHandle.CopyFileMetadataAtIndex(ref options, out FileMetadata? fileMetadata);
+
+                            if (result != Result.Success)
+                            {
+                                tcs.SetException(new Exception($"{nameof(PlayerDataStorageService)} ({nameof(DownloadFile)}): CopyFileMetadataAtIndex returned error. Result code {result}"));
+                                return;
+                            }
+                            else if (fileMetadata != null)
+                            {
+                                if (!string.IsNullOrEmpty(fileMetadata?.Filename))
+                                {
+                                    fileNames.Add(fileMetadata?.Filename);
+                                }
+                            }
```

Comment:
For the sake of clarity, I would excise this code and put it into a `private` method that can be designed statically that is _just responsible_ for doing the copying of file metadata.